core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -654,18 +672,28 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   private def interruptAllChildren(): UIO[Any] =
     if (sendInterruptSignalToAllChildren(_children)) {
       val iterator = _children.iterator()
-
       _children = null
 
-      val body = () => {
-        val next = iterator.next()
+      var curr: Fiber.Runtime[_, _] = null
 
-        if (next != null) next.await(id.location) else Exit.unit
+      //this finds the next operable child fiber and stores it in the `curr` variable
+      def skip() = {
+        curr = null
+        while (iterator.hasNext && (curr eq null)) {
+          curr = iterator.next()
+          if ((curr ne null) && !curr.isAlive())
+            curr = null
+        }
```

Comment:
Just a small performance tip here:

Since the JVM doesn't support nested methods, the scala compiler rewrites them as class methods with extra arguments instead.

When you define `var`s and in the scope outside the nested method and then use them in a nested method (and `Function`s in that matter) the Scala compiler rewrites them as `ObjectRef`s instead which end up in the heap. It's not a big problem, but it means that updates to them will be slower than if they were local variables. In this case, you can avoid some of that overhead, by doing this:

```suggestion
      def skip() = {
        var next = null
        while (iterator.hasNext && (next eq null)) {
          next = iterator.next()
          if ((next ne null) && !next.isAlive())
            next = null
        }
        curr = next
```